### PR TITLE
refactor: [BREAKING] make coffee more extendable

### PR DESCRIPTION
**Pull request by atian25**
- remove 4.x support
- refactor to make `coffee` more extendable
  - move assert to `Rule`
  - `setRule(type, RuleClz)`
  - unknown assert type will throw now
  - `close` will always emit event now
  - ~~`end` will always return promise~~
- use `--require` instead of modify origin file at `beforeScript`
- Introduce `egg-bin` && improve cov
- ~~New Rule for test~~
  - `expect('file', filePath, pattern)` - test file content such as log
  - `expect('file', filePath)` - whether exists
  - `notExpect('file', filePath, pattern)`
  - `notExpect('file', filePath)`

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,7 +1,6 @@
 sudo: false
 language: node_js
 node_js:
-  - '4'
   - '6'
   - '8'
```
Comment (fengmk2) [inline .travis.yml]: add node 10

Comment (atian25) [reply .travis.yml]: done

**Comment by atian25** [comment]
@fengmk2 其实挺想把 `expect('file', filePath, pattern)` 集成进去的，CLI 测试感觉一般都需要验证 log 日志（受你那天的 expectLog 启发）

File: README.md [hunk 2]
```diff
@@ -155,9 +165,32 @@ If you set false, coffee will not generate coverage.json, default: true.
...
+  assert(actual, expected, message) {
+    // do sth
+    return super.assert(content, pattern, message);
+  }
+}
+
+class MyCoffee extends Coffee {
+  constructor(...args) {
+    super(...args);
+    this.RuleMapping.file = FileRule;
```
Comment (fengmk2) [inline README.md]: 是否暴露出一个 `addRule(file, RuleClazz)` api 比较合理？这样后面改动逻辑，或者做 hook 都好做，直接暴露出 RuleMapping 的话，后面重构加功能只能发大版本了。

**Comment by atian25** [comment]
顺便给个权限 `npm owner add atian25 coffee`

**Comment by fengmk2** [comment]
5.0.0

**Comment by fengmk2** [comment]
@atian25 done

**Comment by fengmk2** [comment]
5.0.1，然而还是没法被同步下来。。。

**Comment by fengmk2** [comment]
https://github.com/npm/registry/issues/379

**Pull request merged**